### ISSUE TITLE
Calling dataset groups "groups" (rather than e.g. "categories") is unclear/confusing (terminology)

### ISSUE THREAD

**Issue by Chealer** [ckan/ckan#5331]
### Please describe the expected behaviour
The CKAN interface offers to create something called "groups". This feature appears undocumented, but for many administrators, there is an expectation that groups are for grouping users (like Active Directory groups).

### Please describe the actual behaviour
CKAN groups can in fact group datasets, and not users.

### What steps can be taken to reproduce the issue? 
This is not a behavioral bug.

---

CKAN allows to organize datasets beyond the organization which publishes them. For example, an "Environmental investment" dataset published by the finance ministry can be added to an "Ecology" group, while an "Industrial forecasts" published by the same ministry can be added to an "Industry" group. The best description of groups I could find is in the answer to [this question](https://stackoverflow.com/questions/20963965/whats-the-difference-between-organizations-groups-in-ckan).
<br>

The province I work for created a CKAN instance (Données Québec) which has customized CKAN so that groups are called "catégories" (French for "categories"). In parallel, our province's biggest city has created [a CKAN instance](http://donnees.ville.montreal.qc.ca/group) which has customized CKAN so that groups are called "thématiques" (which I would translate to English as "themes" ([Wiktionary defines  a _thématique_](https://en.wiktionary.org/wiki/th%C3%A9matique) as a "structured group of themes").

Moreover, in just a few months of lurking on the ckan Gitter channel, I have seen 2 people state their intention to rename groups:

- [Abed khooli intends to rename groups to "categories" or "topics".](https://gitter.im/ckan/chat?at=5e841935c090ce1fde35d1ab)
- [olivierdalang intends to rename groups to "projects".](https://gitter.im/ckan/chat?at=5e84b7c155d0d07c2d2e69ab)

I consider the above clear signs that revising the term used to designate groups would be beneficial. A community poll would be required to identify the most consensual alternative, and to what degree users prefer it over "group". The above examples suggest alternatives:

1. "categories"
2. "themes"
3. "topics"
4. "projects"

In my humble opinion, the first 3 alternatives are better, while I would rather keep "groups" than changing for "projects". Another option may be to rename from just "Groups" to "Dataset groups".

There are 2 issues with the current term:

1. The term "group" is very vague. I experimented asking one adult to give me examples of groups, without giving any context, and her answers were all groups of humans. [Wiktionary lists 15 senses for the noun "group"](https://en.wiktionary.org/wiki/group).
2. CKAN's sense for "group" is potentially misleading. One sense of "group" is very likely in CKAN's context: a group of users. The above Wiktionary entry lists it in the computing context. The first 2 sections in [Wikipedia's disambiguation page on "group"](https://en.wikipedia.org/wiki/Group) are about groups of humans, then groups of animals. Even [Wikipedia's article about groups in the computing context](https://en.wikipedia.org/wiki/Group_(computing)) starts with "In computing, the term group generally refers to a grouping of users."

It is non-trivial to find an equally short term which would have no vagueness at all, but I believe "categories" would be much better. I was involved for years in a CMS project which has a similar concept, called "categories", and even though I was involved orders of magnitude more in that project and the product has many more instances than CKAN, I do not remember anyone ever customizing the term, nor even discussing that. The categories feature in that project is optional, but it is widely used.

<br>
I am afraid it is quite late in CKAN's development for such a change, but I see there was already a similar rename ("packages" to "datasets"), and I think this one might be equally important. I am merely proposing to rename in the interface (as was done for datasets); renaming in code/databases would surely be excessively costly.

Abed khooli suggested providing a configuration option for this. If there are several deployments who want to stick to "groups", it might be possible to give the choice between 2 terms, but that would be costly.

Meanwhile, I am planning to write a CKAN 2.8 extension to ease customization. I would write a .pot file and a French .po overriding the standard translations to use "catégories" instead.

**Comment by Chealer** [comment]
Regarding alternatives, one interesting observation is that the equivalent concept in Wordpress (the most used CMS) is [called "categories"](https://wordpress.com/support/posts/categories/).

**Comment by smotornyuk** [comment]
The most important thing here, that `group` in relation to **Organization** is something similar to `package` in relation to **Dataset** in CKAN.
Package - is a base entity for metadata aggregation. this name has meaning inside codebase but is not appropriate for UI. That's why CKAN doesn't use label/name **Package**. But there is one mistake in your observations(or, probably you are absolutely right, but you are talking about old CKAN version. Right now things are a bit different in CKAN) - **Dataset** is not a name/label for `package`. **Dataset** is a label for `package[type=dataset]`. When you are creating packages with different type, they will be using different labels(once again, I'm talking about master and, probably, 2.8 branch - there were a bit different behavior in previous versions).

The **should** be true for `groups`(spoiler: no, right now groups are not working in this way). `group` is an internal representation of user/package aggregating entity. **Organization** - is custom `group[type=organization]`. If you want to create **Category**, **Theme**, **Project** - you should create a `group` with custom type. 
Yes, it possible to use something else by default instead of **Group**. And if it only a question of popularity, then we would definitely agree on switching to **Category**/**Theme**(in my company, we are using **Category** a lot). But it means that we'll introduce another breaking change to all the portal that already using CKAN. So, it's unlikely, that **Group** name will be changed. If you were using default behavior - it will be unchanged. If you have custom code for groups - it still works.

And now the worst part. It would be cool if you just need to create a custom group type and redefine few blocks in templates(like navigation in the header). But currently, labels for **Group** are hardcoded in templates, no matter what is actual `type` of current `group`. Recently I've done some [fixes for `package.type` ](https://github.com/ckan/ckan/pull/5189) usage in templates and URL and now `package` works just as I've described above. And now I'm going to do the same for `group` - after that, you'll be able to use custom `group` types with minimal efforts. But this change won't be backported to CKAN v2.8 and bellow. And, if I won't be fast enough, there is a chance not to get it into v2.9.

**Comment by Chealer** [comment]
Thank you very much for the exhaustive reply and for your (other) work @smotornyuk 

I find your comparison with packages and datasets a bit surprising, considering a couple notes I saw on the topic, such as the following [from the User guide](https://docs.ckan.org/en/2.8/user-guide.html):


> Note
> 
> On early CKAN versions, datasets were called “packages” and this name has stuck in some places, specially internally and on API calls. Package has exactly the same meaning as “dataset”.

But I have no problem to believe you that groups were designed as a concept more general than dataset groups (although I am wondering how custom group types would be created).

I understand the concern about changing the term in labels for something more specific. Indeed, with the situation you explained, that would complicate adding new custom group types.

However, the user interface should not have to suffer from implementation details. An application storing all its objects in a single database table called "objects" does not make it right for that application to call all objects just "objects" regardless of their type. If the application stores pages and files, it should still present these as "Pages" and "Files", not just objects. And it should also list each type on its own page (unless it makes sense to present several types together).

I think there are 2 ways to improve the situation:
1. Change the generic labels to add the type. For example, "My groups" could become "My groups of datasets".
1. Add specific labels for the dataset type of groups. For example, "My groups" could become "My categories" when the type is dataset.

**Comment by Chealer** [comment]
> We cannot only replace **Group** with **Theme** in all the templates, or suggest config option that will be control used term, because it will make harder to use custom groups - either they will be using the same label, or they will be handled in a special way. Conditional branches increase cyclomatic complexity and we trying to work with any entity in the exact same way, so relying on `group.type` is the safest option.

Replacing "Group" with (for example) "Theme" is not the greatest solution, but I am afraid that the implementation complexity does need to be increased. That should not decrease safety, but it does add conditions and translatable strings. If we were discussing a term that only sysadmins have to deal with when they install CKAN, that wouldn't be a big problem. But CKAN should not impose vague terminology in a main concept exposed to end-users just to save some code.

**Comment by smotornyuk** [comment]
No, increasing the complexity won't add a translatable string. The custom group type will do it. No matter which term we will use - it won't express the idea, because `group` is a low-level entity, that was accidentally exposed to end-user at some moment in the past. **Organization** is a group with custom type and using it makes sense. **Group** is a base group and it shouldn't be used directly - but, because it's already there, removing it is quite controversial. Everyone, who doesn't agree with the **Group** should use **custom** group(once against, right now custom groups require additional work).

| feature | Configurable translation of **Group** | custom group types |
| - | - | - |
| Correct name in UI without any custom code, only with config | :heavy_check_mark:   | :heavy_multiplication_x:   |
| Suitable for new portals | :heavy_check_mark: | :heavy_check_mark: |
| Suitable for old portals with custom code | :heavy_multiplication_x: | :heavy_check_mark: |
| Allows multiple group types per portal | :heavy_multiplication_x: | :heavy_check_mark: |
| introduce ambiguity between implementation and UI | :heavy_check_mark: | :heavy_multiplication_x: |

What I'm trying to say - using the configurable label for **Group** is a less flexible and less extendible solution, that fixes only one problem and makes it harder to fix another issue. 

As for this:
> But CKAN should not impose vague terminology in the main concept exposed to end-users just to save some code.
- absolutely agree. But you hardly ever find a better term. All suggested options(except **Project**) only focus on dataset aggregation(which should be done by tags), ignoring the fact, that `group` is rather a collection of users with specific access to the aggregated datasets. So, concrete terms won't solve problem either. because **group shouldn't be used directly at all**.

**Comment by smotornyuk** [comment]
> On early CKAN versions, datasets were called “packages” and this name has stuck in some places, specially internally and on API calls. Package has exactly the same meaning as “dataset”.

CKAN is still evolving and now this sentence is not so correct. It just simplifies things, but actually `package` is more like `object` in your comment, while `dataset` closer to **File** or **Page**. As I understand, you are quite familiar with Wordpress - there is the similar situation there - it using a `post` entity for representing different entities, not only **Post** but also some custom **Article**, **Item**, **Navigation** and a lot of user-introduced content types. Initially, when Wordpress was used as blog-focused cms, using `post` seemed quite right. But with time, as custom entities were introduced, devs had to make `post` configurable.

> Change the generic labels to add the type. For example, "My groups" could become "My groups of datasets".
This one is a perfect solution - it would drastically decrease ambiguity

> Add specific labels for the dataset type of groups. For example, "My groups" could become "My categories" when the type is dataset.
This is not so good. It will increase cyclomatic complexity(note, **cyclomatic complexity**(bad) !== **complexity**(ok)). And this also can be handled by custom dataset types.

Generally, I like your ideas - making things configurable is a best option in most cases. The only detail, I do not agree with - using **custom label in specific case**. No, there should be no specific cases. How about following: 
* Dev can create custom group type with a custom label
* there is a config option that sets this custom group type as default, replacing current **Group**
In such a way we will apply all your suggestion but avoid this special case, that I hate so much :)

Yes, as for custom group types - generally they can be created in the same way, you are creating custom package type. The simplest option - ckanext-scheming(it also possible without any additional extension, but I just love ckanext-scheming):
* install [ckanext-scheming](https://github.com/ckan/ckanext-scheming)
* copy example of schema for [custom group](https://github.com/ckan/ckanext-scheming/blob/master/ckanext/scheming/custom_group_with_status.json) into your extension and adapt fields and `group_type` field for your need
* add `scheming_groups` to the list of enabled plugins
* add your custom schema - `scheming.group_schemas = ckanext.xxx:custom_group.json`
* start server and go to `http://my-local-ckan/<CUSTOM_GROUP_TYPE_BY_SCHEMA>`
That's it

**Comment by Zharktas** [comment]
We did a extension few years ago https://github.com/6aika/ckanext-collection which in essence replicates core groups with different group_type. We did this because we wanted to use groups for categories as it was common convention in instances. 

Replacing groups with something else in .pot file  is usually the easiest way to implement something like categories as group_types are not fully supported in ckan.

Replacing them in .pot file only affect to user interface though, the api still uses groups in everything. Changing them in the api would break every single instance which keeps the ckan updated and should not be done lightly.

**Comment by smotornyuk** [comment]
I think it makes a perfect sense to use this extension - it solves the problem with UI and doesn't break existing features. 
As for the support of `group_types`, yes, I've also mentioned above that right now group types do not provide a full set of required features. But I'm working on it and will do my best in order to get fully-featured group types into v2.9

**Comment by Chealer** [comment]
Thank you @smotornyuk 
> Generally, I like your ideas - making things configurable is a best option in most cases. The only detail, I do not agree with - using **custom label in specific case**. No, there should be no specific cases. How about following:
> 
>     * Dev can create custom group type with a custom label
> 
>     * there is a config option that sets this custom group type as default, replacing current **Group**
>       In such a way we will apply all your suggestion but avoid this special case, that I hate so much :)

When you write "Dev", are you referring to a developer of CKAN?

**Comment by Chealer** [comment]
> As for this:
> 
> > But CKAN should not impose vague terminology in the main concept exposed to end-users just to save some code.
> 
>     * absolutely agree. But you hardly ever find a better term. All suggested options(except **Project**) only focus on dataset aggregation(which should be done by tags), ignoring the fact, that `group` is rather a collection of users with specific access to the aggregated datasets. So, concrete terms won't solve problem either. because **group shouldn't be used directly at all**.

Datasets can be organized with tags, but tags have limitations. The CMS I know best offers 2 different and independant ways to organize content, just like CKAN:

1. Tags (or "freetags"), equivalent to CKAN's tags, the "organic" way
2. Categories. equivalent to CKAN's dataset groups, the organized way

None of these features is better and they are often used in combination.

I do not consider CKAN groups as collection of users. They are associated with users, but what they basically are is collections of datasets.

**Comment by smotornyuk** [comment]
> When you write "Dev", are you referring to a developer of CKAN?

No, to an extension's developer. For example, you can define `group[type=category]` via ckanext-scheming as described above, add something like `ckan.group.default_type = category` to the config file and then all the mentions of **Group**(headings, breadcrumbs, **Groups** tab on dataset page, etc.) will be replaced with **Category**. It won't require much customization from the extension's maintainer, add everything you/ve mentioned and will keep things abstract and extendible.

> The CMS I know best offers 2 different and independent ways to organize content, just like CKAN

Are you sure, that those two ways are really different? I was working only with Drupal and Wordpress, but in both of them, categories(or whatever you call them) are implemented just like custom tags. The same can be done in CKAN - you can create custom tag vocabulary, apply terms from it only to a particular field of the dataset(**Category** field, for example), make a separate page that will show dataset's from particular **Category**(tag under the hood). You can implement anything in UI and no one will ever think, that **Category** is just a `tag`.
 As long as there is no any user management involved(roles, permissions) - you don't need a group. That's why you are wrong, saying
> I do not consider CKAN groups as collection of users. They are associated with users, but what they basically are is collections of datasets.

Tags(not only tag, anything that can be used as search facet, to be honest) are the things that create collection of dataset. Gropus ARE NOT just collections of datasets, they have deeper meaning.

**Comment by Chealer** [comment]
Thanks again @smotornyuk ,

> 
> 
> > When you write "Dev", are you referring to a developer of CKAN?
> 
> No, to an extension's developer. For example, you can define `group[type=category]` via ckanext-scheming inside your local extension, as described above, add something like `ckan.group.default_type = category` to the config file and then all the mentions of **Group**(headings, breadcrumbs, **Groups** tab on dataset page, etc.) will be replaced with **Category**. It won't require much customization from the extension's maintainer, add everything you/ve mentioned and will keep things abstract and extendible.

I see. In that case though, I do not think an extension should be required to have clearly named "categories". I think stock CKAN should come with clear terminology.
 
> > The CMS I know best offers 2 different and independent ways to organize content, just like CKAN
> 
> Are you sure, that those two ways are really different? 

Yes, they can be enabled or disabled independently, and in the back-end they live in independent tables.

> The same can be done in CKAN - you can create custom tag vocabulary, apply terms from it only to a particular field of the dataset(**Category** field, for example), make a separate page that will show datasets from particular **Category**(tag under the hood).

For what it's worth, I do not understand what "apply terms from it only to a particular field of the dataset" means.

> You can implement anything in UI and no one will ever think, that **Category** is just a `tag`.
> As long as there is no user management involved(roles, permissions) - you don't need a group. That's why you are wrong, saying
> 
> > I do not consider CKAN groups as collection of users. They are associated with users, but what they basically are is collections of datasets.
> 
> Tags(not only tag, anything that can be used as search facet, to be honest) are the things that create a collection of the dataset. Groups ARE NOT just collections of datasets, they have a deeper meaning.

Assuming you are discussing the implementation, you may be right that in code or in the database they look like primarily a collection of users. All I meant to say is that for users, dataset groups are a collection of datasets (the role they play is to allow organization of datasets).

> PS there are a lot of such vague spots in docs. It would be really cool if you can create PRs with updates to the doc, from your point of view. It will make documentation more up-to-date and things will become more clear. You are really good at digging into the rationale and have a fresh look at things - CKAN tech team will really appreciate your help.

Frankly, with everything I learned just thanks to your comments in this ticket, I see my understanding of CKAN is still basic and I shouldn't get into deep topics at this point, but thank you.

**Comment by wardi** [comment]
Another light-weight way to group by "topic" or "category" is to use multiple choice fields in your metadata schema https://github.com/ckan/ckanext-scheming#choices then present those choices as facets for your search, e.g. "Portal Type" "Collection Type" "Jurisdiction" etc. on https://open.canada.ca/data/dataset

**Comment by Chealer** [comment]
Thanks @wardi - I don't know if you were addressing me in particular, but the site I work on also uses ckanext-scheming, and it has a "Resource Type" field equivalent to the Federal portal's Collection Type facet, also using a choice field.

By the way, the same CMS I was talking about has a category tree. That makes it possible to create a "Resource Type" category at the root, which has various resource type subcategories as children, so such facets can be created without using an extension. However, I guess that pollutes the category tree with fake categories.

**Comment by smotornyuk** [comment]
> I think stock CKAN should come with clear terminology.

Yep, after all those comments, I can see the problem from your point of view, and even have few ideas about how to introduce such a feature in CKAN. I'll ping you when PR is ready, so you can check it and make some additional suggestions

> apply terms from it only to a particular field

There is an option in CKAN to create independent vocabularies of controlled tags(i.e, you can add, remove, list tags from particular vocabulary). For example, one can create **Keyword**, **Frequency**, **Theme**, and other vocabularies. Each vocabulary will contain an independent set of terms. For every dataset field, one can specify a validator, which allows using terms only from one particular vocabulary. This field can have either multiple values(like the current implementation of the **tags** field) or a single value. After that, you'll be able to filter by this special tag-field just like [by any other facet](https://demo.ckan.org/dataset?groups=economics). And the only thing you need now - implementing the [custom route, that will show the only dataset with this special-tag-field-filter applied](https://demo.ckan.org/group/economics). And from the end-user perspective, there will be no difference from the behavior of **Groups**. 

> Frankly, with everything I learned just thanks to your comments in this ticket, I see my understanding of CKAN is still basic and I shouldn't get into deep topics at this point, but thank you.

You are just looking at "black boxed" CKAN and expecting some functionality. That's the best part - you know what people want to have and what is still unclear. I don't see a lot of problems(or don't think something is a problem, because I've used to it) and thus I cannot fix, make things better, etc. I'm only adding features that I like to have or updating code, when it's ideologically incorrect. You can help in identifying problems in UX, maintenance, setup, etc. Don't underestimate your impact just because you hadn't yet dig deep into CKAN core

**Comment by Chealer** [comment]
We created [a CKAN extension to work around this issue](https://github.com/Chealer/ckanext-categroupy).

**Comment by smotornyuk** [comment]
Don't you mind if I'll mention your extension if this problem will be raised in the future? i mean are there any objections, if I recommend to use it?

**Comment by Chealer** [comment]
Not at all @smotornyuk it just needs realization that it's a quick thing which will require work in most cases (creation of a .po file for all languages other than French at this time).

**Comment by saikarthikp9** [comment]
[ckanext-customizer](https://github.com/saikarthikp9/ckanext-customizer)
Use this extension to replace the terms "group" and "organization" in English. It is a hacky solution. PR/requests are welcome!

**Comment by Chealer** [comment]
Thank you @saikarthikp9 but since this must be indeed quite hacky, I suggest to extend your README to refer those only interested in renaming "groups" to ckanext-categroupy.
